examples/provide.ram/receiver.ram.cpp

```diff
@@ -0,0 +1,43 @@
+#include <eosiolib/eosio.hpp>
+#include <eosiolib/print.hpp>
+#include <eosiolib/bandwith.hpp>
+
+
+struct sharedval{
+    uint64_t id;
+    std::string string;
+
+    uint64_t primary_key() const {return id;}
+};
+
+using valtable = eosio::multi_index<"sharedval"_n, sharedval>;
+
+class [[eosio::contract]] receiver : public eosio::contract {
+public:
+    receiver( eosio::name receiver, eosio::name code, eosio::datastream<const char*> ds) : eosio::contract(receiver, code, ds), values(_self, _self.value) {}
+
+    [[eosio::action]] void write() {
+        eosio::print("Write values to ram provider: ", get_ram_provider(get_self()));
```

Comment:
May be pass ram payer to the function instead of using name of contract?

Reply:
It's need to implement next methods:
`writeself` - smart-contract write data to self memory (don't call `get_ram_provider`)
`writeuser(eosio::name user)` - smart-contract write data to user memory. Smart contract call `get_ram_provider` for get actual ram-payer. Also this method must check authority `require_auto(user)`